### DISCUSSION TITLE
Incomplete export of the libraries breaking compiliation for other libraries

### DISCUSSION THREAD

**Discussion by stevefan1999-personal** [microsoft/vcpkg#21202] [Q&A]
**Describe the bug**
Ditto as title

**Environment**
- OS: Windows 10 amd64
- Compiler: VS2019/MSVC 14.1

**To Reproduce**
1. Install libtorrent & qt & various boost libraries via vcpkg
2. Clone [qbittorrent/qBittorrent](https://github.com/qbittorrent/qBittorrent) and target the CMake build to use vcpkg toolchain
3. Build the generated project using VS2019 under any mode

**Expected behavior**
It should compile and produce executable artifacts

**Failure logs** 
Except it spams "nope, I can't find some symbols, despite they should have had been exported"

**Additional context**
Proof of it being an incompletely exported library: Using my fully exported library to fool the compiler but keep the vcpkg originally built DLL in the binary directory, I was getting "entrypoint not found" error, which means the originally built DLL is also (mysteriously) missing this symbol, no wonder why the compiler is complaining back in the IDE.
![](https://i.imgur.com/BPuvxfi.png)
Thus, I was able to get it working by manually compiling libtorrent with CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS and shimming the DLL with it with this fully exported one.
However, when I was about to make a patch regarding this, I see that in the [guideline](https://vcpkg.readthedocs.io/en/latest/maintainers/maintainer-guide/#do-not-add-cmake_windows_export_all_symbols) that adding CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS is disallowed:
> Do not add CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS

>Unless the author of the library is already using it, we should not use this CMake functionality because it interacts poorly with C++ templates and breaks certain compiler features. Libraries that don't provide a .def file and do not use __declspec() declarations simply do not support shared builds for Windows and should be marked as such with vcpkg_check_linkage(ONLY_STATIC_LIBRARY).

So, we had a dilemma here: We cannot build packages with CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS, yet (maybe the one and only way) to make it work is to use this option.
Yet, libtorrent is impossible to build in static linkage as it creates a huge binary of over 100MB, Nobody wants to wait for half an hour to let it build.
Finally, I suspect there's something wrong with LTCG, or why this would happen will remain a mystery.

**Comment by Neumann-A** [comment]
> Yet, libtorrent is impossible to build in static linkage as it creates a huge binary of over 100MB, Nobody wants to wait for half an hour to let it build.

>Install libtorrent & qt & various boost libraries via vcpkg

I feel like this is a contradiction in itself. qt5-base already takes around 30 min to build.... 
For me it seems more like the compiler flag /Z7 is the real problem of the big binaries.
